### PR TITLE
Adding scan feature with columnFamily and TimeRange

### PR DESCRIPTION
**Pull request by rahulKQL**
Address #2184

This commit is meant to extend the table scan with column family & TimeRange. The user would now be able to scan multiple combinations of CF+TimeRange.

Note: Have cross-checked with Hbase MiniCluster, I believe this filter would provide the same behavior. Also verified with both the integration tests.

### CODE REVIEW

**Comment by rahulKQL** [comment]
@igorbernstein2 @kolea2 Please have a look.

File: bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/TestScan.java [hunk 1]
```diff
@@ -16,13 +16,17 @@
 package com.google.cloud.bigtable.hbase;
 
 import static com.google.cloud.bigtable.hbase.test_env.SharedTestEnvRule.COLUMN_FAMILY;
+import static com.google.cloud.bigtable.hbase.test_env.SharedTestEnvRule.COLUMN_FAMILY2;
+import static org.junit.Assert.assertTrue;
 
 import com.google.protobuf.ByteString;
 import java.io.IOException;
+import java.sql.Date;
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/TestScan.java]: sql date?

File: bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/TestScan.java [hunk 2]
```diff
@@ -305,4 +309,104 @@ public void testStartEndEquals() throws IOException {
...
+      Put put = new Put(rowKeys[rowIndex]);
+      for (int colFamily = 0; colFamily < columnFamilies.length; colFamily++) {
+        for (int qualifierIndex = 0; qualifierIndex < numValuesPerRow; qualifierIndex++) {
+          long time = Date.valueOf(dateRange[colFamily][qualifierIndex]).getTime();
+          put.addColumn(
+              columnFamilies[colFamily], quals[qualifierIndex], time, values[qualifierIndex]);
+        }
+      }
+      puts.add(put);
+    }
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/TestScan.java]: Please clean up the data population. Its very hard to follow there are bunch of parallel data structures that have implicit relationships to each other:
- data ranges are coupled with qualifiers
- why do the rows have to contiguous?
-

Comment (igorbernstein2) [reply bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/TestScan.java]: Also, if you break this up into smaller tests, you won't have to create so much test data

File: bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/TestScan.java [hunk 3]
```diff
@@ -305,4 +309,104 @@ public void testStartEndEquals() throws IOException {
...
+        for (Cell cell : result.rawCells()) {
+          if (Arrays.equals(cell.getFamilyArray(), columnFamilies[0])) {
+
+            long timestamp = cell.getTimestamp();
+            assertTrue("Timestamp is too low", timestamp >= rangeStart);
+            assertTrue("Timestamp is too high", timestamp < rangeEnd);
+          } else {
+            long timestamp = cell.getTimestamp();
+            assertTrue("Timestamp is too low", timestamp >= secRangeStart);
+            assertTrue("Timestamp is too high", timestamp < secRangeEnd);
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/TestScan.java]: Please make this more declarative.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/ScanAdapter.java [hunk 4]
```diff
@@ -261,4 +265,22 @@ private long convertUnits(long hbaseUnits) {
     }
     return Optional.<Filters.Filter>of(interleave);
   }
+
+  private Optional<Filters.Filter> createColFamilyTimeRange(Scan scan) {
+    if (scan.getColumnFamilyTimeRange().isEmpty()) {
+      return Optional.absent();
+    }
+    // Builds filter of the form
+    // ("family1" | "rangeStart, rangeEnd") + ("family2" | "rangeStart2, rangeEnd2")
```
Comment (igorbernstein2) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/ScanAdapter.java]: This notation is confusing `|` usually means OR. It would be easier to read if you use `|` for interleave and `&` for intersection

File: bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/TestScan.java (hunk 1 above)
Comment (rahulKQL) [reply bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/TestScan.java]: I am sorry, I may have used `sql.Date` for `Date#valueOf`, I have now removed this import.

File: bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/TestScan.java (hunk 2 above)
Comment (rahulKQL) [reply bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/TestScan.java]: Thanks for the suggestion, Have now break it up in two test case.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/ScanAdapter.java (hunk 4 above)
Comment (rahulKQL) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/ScanAdapter.java]: I have addressed this comment, please have a look.
